### PR TITLE
[Patterns] Redesign resolution of patterns to follow natural visitation 

### PR DESCRIPTION
**Pull request by srikanth-sankaran**
## Author checklist

- [ ] I have thoroughly tested my changes
- [x] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [x] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

**Review by mpalat — APPROVED** [review]
+1

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/ExpressionContext.java [hunk 1]
```diff
@@ -66,6 +66,20 @@ public boolean definesTargetType() {
 		}
 	},
 
+	/** Instanceof context: potential poly-expressions are: None
+	   Context induced by: InstanceOfExpression.
+	*/
+	INSTANCEOF_CONTEXT {
```
[the remark this answers is not in the document]
Comment (srikanth-sankaran) [reply org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/ExpressionContext.java:72]: > Srikanth - is this context as per spec? [ref: https://docs.oracle.com/javase/specs/jls/se21/html/jls-5.html]

Not really, The key parts are "potential poly-expressions are: None" and the fact that `definesTargetType()` returns `false` - so this is just a variant of `VANILLA_CONTEXT` - just makes it more readable.

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/TypePattern.java [hunk 2]
```diff
@@ -142,7 +138,7 @@ protected boolean isPatternTypeCompatible(TypeBinding other, BlockScope scope) {
 				scope.problemReporter().incompatiblePatternType(this, other, patternType);
 				return false;
 			}
-		} else if (!checkCastTypesCompatibility(scope, other, patternType, this.expression, true)) {
+		} else if (!checkCastTypesCompatibility(scope, other, patternType, null, true)) {
```
[the remark this answers is not in the document]
Comment (srikanth-sankaran) [reply org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/TypePattern.java:141]: > @srikanth-sankaran can you please explain this change?

In conversation with @jarthana this came up - whether the tunneling of `org.eclipse.jdt.internal.compiler.ast.TypePattern.expression` is really called for by something in spec - we concluded based on experimentation it is not.

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/RecordPattern.java [hunk 3]
```diff
@@ -113,85 +115,92 @@ public boolean coversType(TypeBinding t) {
...
-			// check whether the give type reference is a record
-			// check whether a raw type is being used in pattern types
-			// check whether the pattern signature matches that of the record declaration
-			if (!this.resolvedType.isRecord()) {
-				scope.problemReporter().unexpectedTypeinRecordPattern(this.resolvedType, this.type);
-				return this.resolvedType;
+		this.type.bits |= ASTNode.IgnoreRawTypeCheck;
+		this.resolvedType = this.type.resolveType(scope);
+
+		if (!this.resolvedType.isRecord()) {
```
Comment (stephan-herrmann) [inline org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/RecordPattern.java:153]: @srikanth-sankaran  prior to this change, this block sat in `resolveTypeWithBindings` where the case of `this.resolvedType == null` had an early exit prior to it. It's a long shot but could this change be the cause of the NPE reported in https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2399#issuecomment-2315632390 where _some call_ `this.resolvedType.isRecord()` _somewhere_ blew up.

**Pull request merged**